tests/puppeteer/test/linkchecker.js

```diff
@@ -0,0 +1,160 @@
+require('events').EventEmitter.defaultMaxListeners = Infinity;
+const puppeteer = require('puppeteer');
+const fs = require('fs');
+
+const reportPath = './reports/';
+// maybe add tracing option : https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-tracing
+const URL_FO = process.env.URL_FO || 'http://localhost:8080/';
+const URL_BO = process.env.URL_BO || 'http://localhost:8080/admin-dev/';
+const EMAIL = process.env.LOGIN || 'demo@prestashop.com';
+const PASSWD = process.env.PASSWD || 'prestashop_demo';
+const HEADLESS = process.env.HEADLESS || true;
+let responses = [];
+const numberRequests = {BO: 0, FO: 0};
+const selectorBO = 'nav.nav-bar.d-none.d-md-block ul li a.link[href]';
+const selectorFO = 'a[href]';
+let office = null;
+let curHref = null;
+const object = {
+  stats: {
+    start: new Date().getTime(),
+    end: null,
+  },
+  BO: {
+    totalLinks: 0,
+    totalRequests: 0,
+    passed: [],
+    failed: [],
+  },
+  FO: {
+    totalLinks: 0,
+    totalRequests: 0,
+    passed: [],
+    failed: [],
+  },
+};
+
+const getAllUrls = async (page, selector) => {
+  const listUrls = await page.$$eval(selector, as => as.map(a => a.href));
+  // remove perfect duplicates
+  await listUrls.filter((elem, pos) => listUrls.indexOf(elem) === pos);
+  return listUrls;
+};
+
+const interceptRequestAndResponse = async (page) => {
+  await page.setRequestInterception(true);
+  await page.on('request', (request) => {
+    request.continue();
+  });
+  await page.on('response', (response) => {
+    const request = response.request();
+    const url = request.url();
+    const status = response.status().toString();
+    numberRequests[office] += 1;
+    if (status.startsWith('4') || status.startsWith('5')) {
+      console.log(` !! Failed response : ${url} (${status})`);
+      object[office].failed.push({page: curHref, URL: url, status});
+    } else {
+      responses.push({url, status});
+    }
+  });
+};
+
+const checkStatusUrls = async (page, hrefs) => {
+  // check urls
+  let i = 1;
+  // eslint-disable-next-line
+  for (const href of hrefs) {
+    curHref = href;
+    object[office].passed.push({url: href, date: new Date().getTime(), responses});
+
+    if (href.includes(URL_FO)) {
+      // eslint-disable-next-line
+      await page.goto(href, {waitUntil: 'domcontentloaded'});
+      outputSameLine(` - ${i}/${hrefs.length} checked (${href})`);
+    }
+
+    responses = [];
```

Comment:
Any reason to clear responses here?

Reply:
`responses` is the array used to store responses from requests in a page, so we need to clear it after every link crawl. The problem I faced here is that requests logging is made in an event so it's not possible (to my knowledge) to natively link requests to an href directly.
I'd be happy to discuss this if you think there is a workaround.